state/factory/workingset_test.go

```diff
@@ -0,0 +1,138 @@
+package factory
+
+import (
+	"context"
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+
+	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/iotexproject/iotex-core/action"
+	"github.com/iotexproject/iotex-core/action/protocol"
+	"github.com/iotexproject/iotex-core/blockchain/block"
+	"github.com/iotexproject/iotex-core/config"
+	"github.com/iotexproject/iotex-core/pkg/unit"
+	"github.com/iotexproject/iotex-core/test/identityset"
+	"github.com/iotexproject/iotex-core/testutil"
+)
+
+func TestWriteReadView(t *testing.T) {
+	require := require.New(t)
+	sf, err := NewFactory(config.Default, InMemTrieOption())
+	require.NoError(err)
+	ctx := protocol.WithBlockchainCtx(
+		protocol.WithRegistry(context.Background(), protocol.NewRegistry()),
+		protocol.BlockchainCtx{Genesis: config.Default.Genesis},
+	)
+	require.NoError(sf.Start(ctx))
+	ws, err := sf.(workingSetCreator).newWorkingSet(ctx, 1)
+
+	var retrdf interface{}
+	var retErr error
+	retrdf, retErr = ws.ReadView("")
+	require.Equal(protocol.ErrNoName, errors.Cause(retErr))
+	require.Equal(retrdf, nil)
+
+	retrdf, retErr = ws.Unload("")
+	require.Equal(protocol.ErrNoName, errors.Cause(retErr))
+	require.Equal(retrdf, nil)
+
+	require.False(ws.ProtocolDirty("test"))
+
+	require.NoError(ws.WriteView("test", "test"))
+	retrdf, retErr = ws.ReadView("test")
+	require.NoError(retErr)
+	require.Equal(retrdf, "test")
+
+	require.NoError(ws.Load("test", "test"))
+	retrdf, retErr = ws.Unload("test")
+	require.NoError(retErr)
+	require.Equal(retrdf, "test")
+
+	require.True(ws.ProtocolDirty("test"))
+}
+
+func TestValidateBlock(t *testing.T) {
+	require := require.New(t)
+	sf, err := NewFactory(config.Default, InMemTrieOption())
+	require.NoError(err)
+	gasLimit := testutil.TestGasLimit * 100000
+	zctx := protocol.WithBlockCtx(context.Background(),
+		protocol.BlockCtx{
+			BlockHeight: uint64(1),
+			Producer:    identityset.Address(27),
+			GasLimit:    gasLimit,
+		})
+	zctx = protocol.WithBlockchainCtx(
+		zctx,
+		protocol.BlockchainCtx{Genesis: config.Default.Genesis},
+	)
+	require.NoError(sf.Start(zctx))
+
+	digestHash := hash.Hash256b([]byte{65, 99, 99, 111, 117, 110, 116, 99, 117, 114, 114,
+		101, 110, 116, 72, 101, 105, 103, 104, 116, 1, 0, 0, 0, 0, 0, 0, 0})
+
+	t.Run("normal call", func(t *testing.T) {
+		blk := makeBlock(t, 1, hash.ZeroHash256, digestHash)
+		require.NoError(sf.Validate(zctx, blk))
+	})
+
+	t.Run("nonce error", func(t *testing.T) {
+		blk := makeBlock(t, 3, hash.ZeroHash256, digestHash)
+		require.Equal(action.ErrNonce, errors.Cause(sf.Validate(zctx, blk)))
+	})
+
+	t.Run("root hash error", func(t *testing.T) {
+		blk := makeBlock(t, 1, hash.Hash256b([]byte("test")), digestHash)
+		expectedErrors := errors.New("failed to validate block with workingset in factory: " +
+			"Failed to verify receipt root: receipt root hash does not match")
```

Comment:
do same thing:
ErrReceiptRootMismatch = errors.New("receipt root hash does not match")

Reply:
and require.Equal(ErrReceiptRootMismatch, ...)